src/connectionbenchmark/connectionbenchmarkmodel.cpp

```diff
@@ -0,0 +1,131 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "connectionbenchmarkitem.h"
+#include "connectionbenchmarkdownload.h"
+#include "connectionbenchmarkmodel.h"
+#include "leakdetector.h"
+#include "logger.h"
+
+#include <QList>
+
+namespace {
+Logger logger(LOG_MODEL, "ConnectionBenchmarkModel");
+}
+
+ConnectionBenchmarkModel::ConnectionBenchmarkModel() {
+  MVPN_COUNT_CTOR(ConnectionBenchmarkModel);
+}
+
+ConnectionBenchmarkModel::~ConnectionBenchmarkModel() {
+  MVPN_COUNT_DTOR(ConnectionBenchmarkModel);
+}
+
+QHash<int, QByteArray> ConnectionBenchmarkModel::roleNames() const {
+  QHash<int, QByteArray> roles;
+  roles[RoleBenchmark] = "benchmark";
+
+  return roles;
+}
+
+int ConnectionBenchmarkModel::rowCount(const QModelIndex&) const {
+  return m_benchmarks.count();
+}
+
+QVariant ConnectionBenchmarkModel::data(const QModelIndex& index,
+                                        int role) const {
+  if (!index.isValid()) {
+    return QVariant();
+  }
+
+  switch (role) {
+    case RoleBenchmark:
+      return QVariant::fromValue(m_benchmarks.at(index.row()));
+    default:
+      return QVariant();
+  }
+}
+
+void ConnectionBenchmarkModel::setState(State state) {
+  logger.debug() << "Set state" << state;
+  m_state = state;
+
+  emit stateChanged();
+}
+
+void ConnectionBenchmarkModel::addResult(
+    ConnectionBenchmarkItem* benchmarkItem) {
+  logger.debug() << "Add benchmark result";
+
+  beginInsertRows(QModelIndex(), m_benchmarks.size(), m_benchmarks.size());
+  m_benchmarks.append(benchmarkItem);
+  endInsertRows();
+}
+
+void ConnectionBenchmarkModel::runNextBenchmark() {
+  if (m_state == StatePing) {
+    logger.debug() << "Run ping benchmark";
+
+    // TODO: Add localized string
+    addResult(new ConnectionBenchmarkItem(
+        "ping", "Ping", "qrc:/nebula/resources/connection-green.svg", 15));
+
+    setState(StateDownload);
+    runNextBenchmark();
+
+  } else if (m_state == StateDownload) {
+    logger.debug() << "Run download benchmark";
```

Comment:
I like that we have states because we check items one by one. We should expose that download value is not ready yet.